test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,505 @@
+local cluster = require('test.luatest_helpers.cluster')
+local g_nspaces = luatest.group('gh-6033-box-promote-demote')
+local g_unconfigured = luatest.group('gh-6033-box-promote-demote-unconfigured')
+local g_spaces = luatest.group('gh-6033-box-promote-demote-with-spaces',
+    {{engine = 'memtx'}, {engine = 'vinyl'}})
+
+local promote_start = function(server)
+    return server:exec(function()
+        local f = require('fiber').new(box.ctl.promote)
+        f:set_joinable(true)
+        return f:id()
+    end)
+end
+
+local demote_start = function(server)
+    return server:exec(function()
+        local f = require('fiber').new(box.ctl.demote)
+        f:set_joinable(true)
+        return f:id()
+    end)
+end
+
+local fiber_join = function(server, fiber)
+    return server:exec(function(fiber)
+        return require('fiber').find(fiber):join()
+    end, {fiber})
+end
+
+local wal_delay_start = function(server,  countdown)
+    if countdown == nil then
+        server:exec(function()
+            box.error.injection.set('ERRINJ_WAL_DELAY', true)
+        end)
+    else
+        server:exec(function(countdown)
+            box.error.injection.set('ERRINJ_WAL_DELAY_COUNTDOWN', countdown)
+        end, {countdown})
+    end
+end
+
+local wal_delay_end = function(server)
+    server:exec(function()
+        box.error.injection.set('ERRINJ_WAL_DELAY', false)
+    end)
+end
+
+local cluster_init = function(g)
+    g.cluster = cluster:new({})
+
+    local engine = 'memtx'
+    if g.params ~= nil and g.params.engine ~= nil then
+        engine = g.params.engine
+    end
+
+    g.box_cfg = {
+        election_mode = 'off',
+        election_timeout = 0.5,
+        read_only = false,
+        replication_timeout = 0.1,
+        replication_synchro_timeout = 5,
+        replication_synchro_quorum = 1,
+        replication = {
+            server.build_instance_uri('server_1'),
+            server.build_instance_uri('server_2'),
+        },
+    }
+
+    g.server_1 = g.cluster:build_and_add_server(
+        {alias = 'server_1', engine = engine, box_cfg = g.box_cfg})
+    g.server_2 = g.cluster:build_and_add_server(
+        {alias = 'server_2', engine = engine, box_cfg = g.box_cfg})
+    g.cluster:start()
+end
+
+local cluster_reinit = function(g)
+    g.server_1:wait_lsn(g.server_2)
+    g.server_2:wait_lsn(g.server_1)
+
+    g.server_1:box_config(g.box_cfg)
+    g.server_2:box_config(g.box_cfg)
+end
+
+g_nspaces.before_all(cluster_init)
+g_spaces.before_all(cluster_init)
+
+g_nspaces.after_all(function(g) g.cluster:drop() end)
+g_spaces.after_all(function(g) g.cluster:drop() end)
+
+g_nspaces.after_each(cluster_reinit)
+g_spaces.after_each(cluster_reinit)
+
+-- Promoting/demoting should succeed if server is not configured.
+g_unconfigured.test_unconfigured = function()
+    local ok, err = pcall(box.ctl.promote)
+    luatest.assert(ok, string.format(
+        'error while promoting unconfigured server: %s', err))
```

Comment:
Message in assertion should say what you expect. Not what you do not expect. Here you assert `ok` but say it is an error. It should be just *"promoting unconfigured server"*. Not an error. If you want to specifically test the error and get its value in case of failure, you can add a line `luatest.assert_equals(err, nil, 'no error')`. Then when fails, it will print the value of `err`. Same in other places.